### PR TITLE
Openid Connect implementation

### PR DESCRIPTION
**Pull request by xerbalind**
It follows Authorization Code Flow using Oauth2.
The **id_token** is not yet signed as it should be, described in [the specification](https://openid.net/specs/openid-connect-core-1_0.html#IDToken). **IMPLEMENTED**
UserInfo does not yet contain the **sub** claim (see [the specification](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)).
The **id** field could be replaced with **sub** at `/current_user` or another endpoint could be added specific for OpenId Connect.

Support for scopes might be handy.

With this pull request gitea is happy :smile: (gitea does not even check the **id_token** signature)

### CODE REVIEW

File: src/controllers/oauth_controller.rs [hunk 1]
```diff
@@ -236,10 +239,22 @@ fn authorization_denied(state: AuthState) -> Redirect {
...
 }
 
+#[derive(Serialize, Debug)]
+pub struct IDToken {
+	sub:      String,
+	iss:      String,
+	aud:      String,
+	exp:      i64,
+	iat:      i64,
+	nickname: String,
```
Comment (rien) [inline src/controllers/oauth_controller.rs:249]: I would use `username` here for consistency throughout the code. Or is there a specific reason you chose `nickname`?

**Review by rien — COMMENTED** [review]
If you want to implement OpenID with JWT completely, you might want to check out the https://github.com/Keats/jsonwebtoken crate, it seems to be the most popular crate for creating and veryfing signed JWT tokens.

I'll take a deeper look into this later, but I think we might want to add another check to see if the `id_token` is implemented correctly.

File: src/controllers/oauth_controller.rs (hunk 1 above)
Comment (xerbalind) [reply src/controllers/oauth_controller.rs:249]: `username` is not a standard claim (see [specification on claims](https://openid.net/specs/openid-connect-basic-1_0.html#StandardClaims)). `preferred_username` is , and will work as alternative.

Comment (rien) [reply src/controllers/oauth_controller.rs:249]: Yes, I think `preferred_username` would be a better fit.

**Review by NuttyShrimp — APPROVED** [review]

**Review by rien — APPROVED** [review]
Looking good!

The OAuth test code is cleaner now as well, thanks.

**Pull request merged**